### ISSUE TITLE
Site performance needs optimisation 

### ISSUE THREAD

**Issue by nIvab** [nIvab/cocktailFinder#1]
**TL;DR:**  Site performance is suffering due to the way that API call on search is handled

The following is from `src/App.js` 
``` 
// tags are the search terms the user inputs 
const [tags, setTags] = useState([])
// searchResults is the complete data retrieved from the API call
const [searchResults, setSearchResults] = useState([]);
//for loading screen 
const [isLoading, setLoading] = useState(true);
```
```
 useEffect(() => {

        // apiSearchCall contains the incomplete data 

        apiSearchCall(tags).then((data) => {
            if (data.drinks === "None Found") {
                setNoDrinks(true);
            } else {
                let idArr = aggregateID(data);
                let drinks = [];
                idArr.forEach((element) => {                                      // <---  ID OF EACH DRINK IN idArr 
                    // apiIndividualCall contains the complete data
                    apiIndividualCall(element).then((dataFinal) => {
                        drinks.push(dataFinal.drinks[0]);                  // <---- drinks ARRAY IS RECONSTRUCTED ON EACH ITERATION
                                                                                                // WHICH IS NOT OPTIMAL 
                        setSearchResults({ drinks });
                    });
                });
                setLoading(false);
            }
        });
    }, [setSearchResults, setLoading, tags]);

 ```
searchResults is then sent to another component `Card.js`

While this does give me the functionality I do need([Here is a video demonstrating it](https://i.imgur.com/DyVHlwk.mp4)), it comes at a hefty performance hit as the searchResults are reset on every iteration of the forEach loop and *I think* each component is re-rendered. In an attempt to optimise this, I have tried the following separately 

 - Lifted the pre-allocation of drinks to outside of the `.then()` and called `setSearchResults` and `setLoading` outside of the `.then()`

- removed `setLoading` and `setSearchResults` and added them another chained `.then()` 

- removed the forEach loop and added it to a seperate helper function which returns what would've been the contents of drinks and passed that to `setSearchResults` 

**All** of these attempts have resulted in [the components not rendering](https://i.imgur.com/zKgiqOp.mp4). After console.log-ing a lot of my code I discovered the **right information is being passed to the card component** and all the logic runs up-until we hit the return statement where nothing is run from there on. 

My gut feeling is that I might be missing something fundamental about how `useEffect()` works, but I am not sure what it might be 

 
If it would help at all I am utilising this API
 
https://www.thecocktaildb.com/api.php

Where the results of the search for ingredients (`apiSearchCall()` in my code) has the following structure  
```
{
"drinks": [
    {
      "strDrink": "Chocolate Beverage",
      "strDrinkThumb": "https://www.thecocktaildb.com/images/media/drink/jbqrhv1487603186.jpg",
      "idDrink": "12732"
    }
  ]
}
```

[Link to StackOverflow question regarding this very same issue](https://stackoverflow.com/questions/65987853/react-optimising-api-call-chain-causes-components-to-not-render?noredirect=1#comment116670967_65987853)

**Comment by gbudjeakp** [comment]
Issue: So I have been looking at the code base and have a question, do you have a reason you opted to use for-loops and forEach rather then using .map() filter and .reduce()? They make your code base easy to read and might just help solve your issue.  Your code is constantly looping and then comparing each iteration to the input rather than just filtering and matching. Also one thing that is constant throughout the code base,  is that each  loop and push, causes the  useEffect to activate which is why it takes a long time for things to render. you only want useEffect to trigger once there is a set array. but since you are looping and pushing, that means that the array that is supposed to be rendered/utilized, is being compiled and useEffect will not load until the loop has been concluded thus the long wait.

Tip:
Try switching from for loops and forEach to map, filter and reduce. map will create a new array containing the results of whatever function you provided. You can then render the result rather than looping through each element in the array before rendering. Try looking them up if you haven't heard of them. With map, it will loop first then provide a new complete array with whatever you modified on it and then the useEffect will trigger a complete array rather than waiting for an array that is building up.

filter method creates a new array with all elements that pass the test implemented by the provided function. So say when you type `Gin`, you only want items containing the word `Gin`. filter will only show the result of items containing the word `Gin`

**Comment by nIvab** [comment]
Hi, thank you for your suggestion! It did end up leading me to the required fix! 

Generally speaking, throughout the codebase I mainly used for loops and forEach as that what made sense to me at the time. However a refactor is probably needed to improve readability. 

As for the fix, I created a new array that was the result of a mapping of idArr to a promise containing the appropriate drink (relative to the id of the drink contained in idArr) then resolved this array of promises using Promise.all 

``` 
useEffect(() => {
        apiSearchCall(tags).then((data) => {
            if (data.drinks === "None Found") {
                setNoDrinks(true);
            } else {
                let idArr = aggregateID(data);
                let drinksPromise = idArr.map((element) => {
                    return apiIndividualCall(element).then((searchData) => {
                        return searchData.drinks[0];
                    });
                });
                Promise.all(drinksPromise).then((drinks) => {
                    setLoading(false);
                    setSearchResults({ drinks });
                });
            }
        });
    }, [setSearchResults, setLoading, tags]);
```
The fix is now live on the site
